x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ServiceAccountService.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.CharArrays;
+import org.elasticsearch.common.hash.MessageDigests;
+import org.elasticsearch.common.io.stream.InputStreamStreamInput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.security.authc.service.ServiceAccount.ServiceAccountId;
+import org.elasticsearch.xpack.security.authc.support.SecurityTokenType;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.security.authc.service.ElasticServiceAccounts.ACCOUNTS;
+
+public class ServiceAccountService {
+
+    public static final String REALM_TYPE = "service_account";
+    public static final String REALM_NAME = "service_account";
+
+    private static final Logger logger = LogManager.getLogger(ServiceAccountService.class);
+    private static final Version VERSION_MINIMUM = Version.V_8_0_0;
+
+    private final ServiceAccountsCredentialStore serviceAccountsCredentialStore;
+
+    public ServiceAccountService(ServiceAccountsCredentialStore serviceAccountsCredentialStore) {
+        this.serviceAccountsCredentialStore = serviceAccountsCredentialStore;
+    }
+
+    public static boolean isServiceAccount(Authentication authentication) {
+        return REALM_TYPE.equals(authentication.getAuthenticatedBy().getType());
+    }
+
+    // {@link org.elasticsearch.xpack.security.authc.TokenService#extractBearerTokenFromHeader extracted} from an HTTP authorization header.
```

Comment:
Should this be part of the javadoc below?

Reply:
Yes, but `TokenService` is not updated yet and this causes compile error. I commented it out with the intention to restore it later.